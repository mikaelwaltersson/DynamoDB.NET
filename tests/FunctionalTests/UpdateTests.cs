using Amazon.DynamoDBv2.Model;
using DynamoDB.Net.Tests.FunctionalTests.Common;
using DynamoDB.Net.Tests.FunctionalTests.Common.Models;
using static DynamoDB.Net.DynamoDBExpressions;

namespace DynamoDB.Net.Tests.FunctionalTests.TestModels;

public class UpdateTests : FunctionalTestSuiteBase
{
    [SeedItem]
    readonly User existingUser1 = 
        new()
        { 
            UserId = new Guid("8670ca59-b316-43b2-a6bf-103825793018"), 
            Username = "johndoe", FullName = "John Doe" 
        };

    [SeedItem]
    readonly UserNotification existingUserNotification1 =
        new()
        { 
            UserId = new Guid("8670ca59-b316-43b2-a6bf-103825793018"), 
            Timestamp = DateTimeOffset.Parse("2024-04-15T08:41:00Z"),
            Message = "Your account will soon be available." 
        };

    [Fact]
    public async Task CanUpdateItemsAsync()
    {
        var user = existingUser1;
        var userNotification = existingUserNotification1;

        user = await DynamoDBClient.UpdateAsync(user, () => Remove(user.FullName));
        userNotification = await DynamoDBClient.UpdateAsync(userNotification, () => Set(userNotification.Message, "<REDACTED>"));
        
        var storedItems = (await GetAllStoredRawItemsAsync()).ToSnapshotFriendlyObject();

        Snapshot.Match(
            new 
            {
                user,
                userNotification,
                storedItems
            });
    }

    [Fact]
    public async Task UpdateFailsIfVersionDoesNotMatchAsync()
    {
        var user = existingUser1;

        await Assert.ThrowsAsync<ConditionalCheckFailedException>(() => 
            DynamoDBClient.UpdateAsync(
                user, 
                () => Remove(user.FullName),
                version: 2));
    }

    [Fact]
    public async Task UpdateItemFailsIfConditionIsNotSatisfiedAsync()
    {
        var user = existingUser1;
        var userNotification = existingUserNotification1;

        await Assert.ThrowsAsync<ConditionalCheckFailedException>(() => 
            DynamoDBClient.UpdateAsync(user, 
            () => Remove(user.FullName), 
            condition: () => user.FullName == null));

        await Assert.ThrowsAsync<ConditionalCheckFailedException>(() =>
            DynamoDBClient.UpdateAsync(
                userNotification, 
                () => Set(userNotification.Message, "<REDACTED>"), 
                condition: () => userNotification.Message.Count() > 400));
    }
}
