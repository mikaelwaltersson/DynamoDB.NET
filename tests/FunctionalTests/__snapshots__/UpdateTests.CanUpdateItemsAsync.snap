{
  "user": {
    "UserId": "8670ca59-b316-43b2-a6bf-103825793018",
    "UserStatus": "Active",
    "Username": "johndoe",
    "FullName": null,
    "ItemVersion": 1
  },
  "userNotification": {
    "UserId": "8670ca59-b316-43b2-a6bf-103825793018",
    "Timestamp": "2024-04-15T08:41:00+00:00",
    "Message": "<REDACTED>"
  },
  "storedItems": [
    {
      "Table": "users",
      "Items": [
        {
          "userStatus": {
            "S": "active"
          },
          "$version": {
            "N": "1"
          },
          "userId": {
            "S": "8670ca59-b316-43b2-a6bf-103825793018"
          },
          "$type": {
            "S": "DynamoDB.Net.Tests.FunctionalTests.Common.Models.User, DynamoDB.Net.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null"
          },
          "username": {
            "S": "johndoe"
          }
        }
      ]
    },
    {
      "Table": "user-notifications",
      "Items": [
        {
          "message": {
            "S": "<REDACTED>"
          },
          "userId": {
            "S": "8670ca59-b316-43b2-a6bf-103825793018"
          },
          "$type": {
            "S": "DynamoDB.Net.Tests.FunctionalTests.Common.Models.UserNotification, DynamoDB.Net.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null"
          },
          "timestamp": {
            "S": "2024-04-15T08:41:00.0000000+00:00"
          }
        }
      ]
    }
  ]
}
